pkgs/build-support/trivial-builders: add runCommandLocal

Profpatsch opened the pull request:
A definition I’ve been copy-pasting everywhere so far, so it’s finally
time to add it to nixpkgs.

I’m using a remote builder for my regular nix builds, so trivial
`runCommand`s which first try a substitution and then copy the inputs
to the builder to run for 0.2s are quite noticable.

If we just always build these, we gain some build time, so let’s make
it easy to switch from remote to local.

<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @Infinisil

Profpatsch committed:
pkgs/build-support/trivial-builders: add runCommandLocal

A definition I’ve been copy-pasting everywhere so far, so it’s finally
time to add it to nixpkgs.

I’m using a remote builder for my regular nix builds, so trivial
`runCommand`s which first try a substitution and then copy the inputs
to the builder to run for 0.2s are quite noticable.

If we just always build these, we gain some build time, so let’s make
it easy to switch from remote to local.

grahamc commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -21,10 +27,26 @@ rec {
+  runCommandNoCCLocal = runCommand' true stdenvNoCC;
+
+  runCommandCC = runCommand' false stdenv;
+  runCommandCCLocal = runCommand' true stdenv;
```
Let's get these documented: https://nixos.org/nixpkgs/manual/#chap-trivial-builders

Profpatsch replied:
It’s worrying to me that the API docs differ between the manual and the docstrings.

grahamc replied:
Go ahead and update them, too :)

Profpatsch replied:
Done.

Profpatsch committed:
doc/builders/trivial-builders: added runCommandLocal docs

The link in the note points to the `id` added in
https://github.com/NixOS/nix/pull/3255, so it might take some time to
start working correctly.

Infinisil commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -2,11 +2,17 @@
+    // (if runLocal then {
+          preferLocalBuild = true;
+          allowSubstitutes = false;
+       } else {})
```
I feel like this should be the other way around, such that `env` can override `preferLocalBuild`. Also could use `optionalAttrs runLocal { ... }` here

Profpatsch replied:
You are absolutely right, see force push.

Tested manually with

```
nix-build -E 'with import ./. {}; runCommandLocal "foo" { x = "y"; preferLocalBuild = false; } "echo $x"'
```

Profpatsch committed:
doc/builders: add ids to the trivial builder definitions

This makes it possible to reference single function definitions,
for pointing people to their exact definition.

Profpatsch committed:
pkgs/build-support/trivial-builders: remove runCommandCCLocal

We shouldn’t force the user to have a C compiler in scope, just
because the derivation is forced to build locally. That can’t be
counted as “lightweight” anymore.

Co-Authored-By: Silvan Mosberger<contact@infinisil.com>
